test/Streamly/Test/Prelude/Serial.hs

```diff
@@ -106,55 +107,258 @@ splitOnSuffixSeq = do
+
+            -- Karp-Rabin
+            intercalateSplitEqId 4
+            intercalateSplitEqIdS 4
+            concatSplitIntercalateEqConcat splitOnSeq_ intercalate 4
+            concatSplitIntercalateEqConcatS splitOnSeq_ IS.intercalate 4
+
+            -- Exclusive case
+            splitIntercalateEqId splitOnSeq_ intercalate
+            splitIntercalateEqIdS splitOnSeq_ IS.intercalate
+
+    describe (desc <> " splitOnSuffixSeq")
+        $ do
+            -- Empty seperator
+            intercalateSplitEqIdNoSepEnd 0
+            intercalateSplitEqIdNoSepEndS 0
+            concatSplitIntercalateEqConcat splitOnSuffixSeq_ intercalateSuffix 0
+            concatSplitIntercalateEqConcatS
+                splitOnSuffixSeq_
+                IS.intercalateSuffix
+                0
+
+            -- Single element seperator
+            intercalateSplitEqIdNoSepEnd 1
+            intercalateSplitEqIdNoSepEndS 1
+            concatSplitIntercalateEqConcat splitOnSuffixSeq_ intercalateSuffix 1
+            concatSplitIntercalateEqConcatS
+                splitOnSuffixSeq_
+                IS.intercalateSuffix
+                1
+
+            -- Shift Or
+            intercalateSplitEqIdNoSepEnd 2
+            intercalateSplitEqIdNoSepEndS 2
+            concatSplitIntercalateEqConcat splitOnSuffixSeq_ intercalateSuffix 2
+            concatSplitIntercalateEqConcatS
+                splitOnSuffixSeq_
+                IS.intercalateSuffix
+                2
+
+            -- Karp-Rabin
+            intercalateSplitEqIdNoSepEnd 4
+            intercalateSplitEqIdNoSepEndS 4
+            concatSplitIntercalateEqConcat splitOnSuffixSeq_ intercalateSuffix 4
+            concatSplitIntercalateEqConcatS
+                splitOnSuffixSeq_
+                IS.intercalateSuffix
+                4
+
+            -- Exclusive case
+            splitIntercalateEqId splitOnSuffixSeq_ intercalateSuffix
+            splitIntercalateEqIdS splitOnSuffixSeq_ IS.intercalateSuffix
+
+    where
+
+    splitOnSeq_ xs ys =
+        S.toList $ IS.splitOnSeq (A.fromList ys) FL.toList (S.fromList xs)
+
+    splitOnSuffixSeq_ xs ys =
+        S.toList $ IS.splitOnSuffixSeq (A.fromList ys) FL.toList (S.fromList xs)
+
+    intercalateSuffix xs yss = intercalate xs yss ++ xs
+
+    nonSepElem :: Gen a
+    nonSepElem = suchThat arbitrary (/= sep)
+
+    listWithSep :: Gen [a]
+    listWithSep = listOf $ frequency [(3, arbitrary), (1, elements [sep])]
+
+    listWithoutSep :: Gen [a]
+    listWithoutSep = vectorOf 4 nonSepElem
+
+    listsWithoutSep :: Gen [[a]]
+    listsWithoutSep = listOf listWithoutSep
+
+    listsWithoutSep1 :: Gen [[a]]
+    listsWithoutSep1 = listOf1 listWithoutSep
+
+    intercalateSplitEqId i =
+        let name =
+                "intercalate . splitOnSeq == id ("
+                    <> show i <> " element separator)"
+         in prop name
+                $ forAll listWithSep
+                $ \xs -> do
```

Comment:
We can move this do block to a where clause to reduce the indentation. Or we can assign the `intercalate` in `listEquals` to a variable so that `listEquals` check becomes a single line.